File: src/clipboard.ts
import { spawnSync } from "node:child_process";

interface ClipboardCommand {
  command: string;
  args: string[];
}

/** The result of an expected clipboard operation. */
export type ClipboardResult<T> =
  | { readonly ok: true; readonly value: T }
  | { readonly ok: false; readonly message: string };

function readCommands(): ClipboardCommand[] {
  if (process.platform === "darwin") return [{ command: "pbpaste", args: [] }];
  if (process.platform === "win32") {
    return [
      {
        command: "powershell.exe",
        args: ["-NoProfile", "-NonInteractive", "-Command", "Get-Clipboard -Raw"],
      },
    ];
  }
  return [
    { command: "wl-paste", args: ["--no-newline"] },
    { command: "xclip", args: ["-selection", "clipboard", "-out"] },
    { command: "xsel", args: ["--clipboard", "--output"] },
  ];
}

function writeCommands(): ClipboardCommand[] {
  if (process.platform === "darwin") return [{ command: "pbcopy", args: [] }];
  if (process.platform === "win32") {
    return [
      {
        command: "powershell.exe",
        args: ["-NoProfile", "-NonInteractive", "-Command", "$input | Set-Clipboard"],
      },
    ];
  }
  return [
    { command: "wl-copy", args: [] },
    { command: "xclip", args: ["-selection", "clipboard", "-in"] },
    { command: "xsel", args: ["--clipboard", "--input"] },
  ];
}

/** Read text from the first clipboard adapter available on the current platform. */
export function readClipboard(): ClipboardResult<string> {
  for (const candidate of readCommands()) {
    const result = spawnSync(candidate.command, candidate.args, {
      encoding: "utf8",
      stdio: ["ignore", "pipe", "ignore"],
      windowsHide: process.platform === "win32",
    });
    if (result.status === 0 && typeof result.stdout === "string") {
      return { ok: true, value: result.stdout };
    }
  }
  return { ok: false, message: "No supported clipboard reader is available" };
}

/** Write text through the first clipboard adapter available on the current platform. */
export function writeClipboard(text: string): ClipboardResult<undefined> {
  for (const candidate of writeCommands()) {
    const result = spawnSync(candidate.command, candidate.args, {
      input: text,
      encoding: "utf8",
      stdio: ["pipe", "ignore", "ignore"],
      windowsHide: process.platform === "win32",
    });
    if (result.status === 0) return { ok: true, value: undefined };
  }
  return { ok: false, message: "No supported clipboard writer is available" };
}

File: rust/src/clipboard.rs
//! Per-platform clipboard adapters.

use std::io::Write;
use std::process::{Command, Stdio};

#[derive(Debug, Clone, PartialEq, Eq)]
struct ClipboardCommand {
    command: &'static str,
    args: &'static [&'static str],
}

fn read_commands() -> Vec<ClipboardCommand> {
    if cfg!(target_os = "macos") {
        return vec![ClipboardCommand {
            command: "pbpaste",
            args: &[],
        }];
    }
    if cfg!(windows) {
        return vec![ClipboardCommand {
            command: "powershell.exe",
            args: &[
                "-NoProfile",
                "-NonInteractive",
                "-Command",
                "Get-Clipboard -Raw",
            ],
        }];
    }
    vec![
        ClipboardCommand {
            command: "wl-paste",
            args: &["--no-newline"],
        },
        ClipboardCommand {
            command: "xclip",
            args: &["-selection", "clipboard", "-out"],
        },
        ClipboardCommand {
            command: "xsel",
            args: &["--clipboard", "--output"],
        },
    ]
}

fn write_commands() -> Vec<ClipboardCommand> {
    if cfg!(target_os = "macos") {
        return vec![ClipboardCommand {
            command: "pbcopy",
            args: &[],
        }];
    }
    if cfg!(windows) {
        return vec![ClipboardCommand {
            command: "powershell.exe",
            args: &[
                "-NoProfile",
                "-NonInteractive",
                "-Command",
                "$input | Set-Clipboard",
            ],
        }];
    }
    vec![
        ClipboardCommand {
            command: "wl-copy",
            args: &[],
        },
        ClipboardCommand {
            command: "xclip",
            args: &["-selection", "clipboard", "-in"],
        },
        ClipboardCommand {
            command: "xsel",
            args: &["--clipboard", "--input"],
        },
    ]
}

fn process(command: &str) -> Command {
    let value = Command::new(command);
    #[cfg(windows)]
    let value = {
        use std::os::windows::process::CommandExt;
        let mut value = value;
        value.creation_flags(0x0800_0000);
        value
    };
    value
}

/// Read text from the first clipboard adapter available on the current platform.
pub fn read_clipboard() -> Result<String, String> {
    for candidate in read_commands() {
        let result = process(candidate.command)
            .args(candidate.args)
            .stdin(Stdio::null())
            .stderr(Stdio::null())
            .output();
        if let Ok(output) = result
            && output.status.success()
        {
            return Ok(String::from_utf8_lossy(&output.stdout).into_owned());
        }
    }
    Err("No supported clipboard reader is available".to_owned())
}

/// Write text through the first clipboard adapter available on the current platform.
pub fn write_clipboard(text: &str) -> Result<(), String> {
    for candidate in write_commands() {
        let spawned = process(candidate.command)
            .args(candidate.args)
            .stdin(Stdio::piped())
            .stdout(Stdio::null())
            .stderr(Stdio::null())
            .spawn();
        let Ok(mut child) = spawned else { continue };
        let wrote = child
            .stdin
            .take()
            .is_some_and(|mut stdin| stdin.write_all(text.as_bytes()).is_ok());
        if wrote && child.wait().is_ok_and(|status| status.success()) {
            return Ok(());
        }
        let _ = child.kill();
        let _ = child.wait();
    }
    Err("No supported clipboard writer is available".to_owned())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn every_platform_has_read_and_write_candidates() {
        assert!(!read_commands().is_empty());
        assert!(!write_commands().is_empty());
    }
}
